src/compiler/checker.ts

```diff
@@ -30539,20 +30539,18 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         // A variable is considered uninitialized when it is possible to analyze the entire control flow graph
         // from declaration to use, and when the variable's declared type doesn't include undefined but the
         // control flow based type does include undefined.
-        if (!isEvolvingArrayOperationTarget(node) && (type === autoType || type === autoArrayType)) {
-            if (flowType === autoType || flowType === autoArrayType) {
-                if (noImplicitAny) {
-                    error(getNameOfDeclaration(declaration), Diagnostics.Variable_0_implicitly_has_type_1_in_some_locations_where_its_type_cannot_be_determined, symbolToString(symbol), typeToString(flowType));
-                    error(node, Diagnostics.Variable_0_implicitly_has_an_1_type, symbolToString(symbol), typeToString(flowType));
-                }
-                return convertAutoToAny(flowType);
-            }
-        }
-        else if (!assumeInitialized && !containsUndefinedType(type) && containsUndefinedType(flowType)) {
+        if (!typeIsAutomatic && !assumeInitialized && !containsUndefinedType(type) && containsUndefinedType(flowType)) {
             error(node, Diagnostics.Variable_0_is_used_before_being_assigned, symbolToString(symbol));
             // Return the declared type to reduce follow-on errors
             return type;
         }
+        if (flowType === autoType || flowType === autoArrayType && !isEvolvingArrayOperationTarget(node)) {

```

Comment:
I simplified this a little bit. We don't need to check if `type` is an auto type - it's sufficient to check `flowType`s. They can't be auto types if the `type` wasn't auto in the first place.

On top of that, `isEvolvingArrayOperationTarget` is only checked now when `flowType === autoArrayType`. This fixes an extra bug ([TS playground](https://www.typescriptlang.org/play/?ts=5.7.0-dev.20240830&ssl=11&ssc=2&pln=1&pc=1#code/PTAEDsHtQUwJzpOBnUheDcIx7AoANjALqPjMvgIwDMA3FgGYCu4AxvgJaTii3hkAUAlKADeWUKDgF6cTkNB5wAc3wALAFxES5CgDp5S5aAC+NQ1iwhYCJMgCE6QLI7uAqEgr4lGg2ZsOXcACYBYVFxSWlhLkhIdVdldx1aKKMTLCA)):
```ts
// no errors 😱
let test13;
function fn1() {
  return { length: test13.length };
}

// errors! 👍
let other13;
function fn2() {
  return { foo: other13.foo };
}
```